addons/l10n_br_pix/models/res_partner_bank.py

```diff
@@ -0,0 +1,111 @@
+                raise ValidationError(
+                    _(
+                        "The mobile number must start with +55, contain a 2 digit territory or state code followed by a 10 digit number."
+                    )
+                )
+
+            regex = r"%(char)s{8}-%(char)s{4}-%(char)s{4}-%(char)s{4}-%(char)s{12}" % {"char": "[a-fA-F0-9]"}
+            if bank.proxy_type == "br_random" and not re.fullmatch(regex, bank.proxy_value):
+                raise ValidationError(
+                    _("%s is not a valid random key. The format looks like this: 71d6c6e1-64ea-4a11-9560-a10870c40ca2")
+                )
+
+    @api.depends("country_code")
+    def _compute_display_qr_setting(self):
+        """Override."""
+        bank_br = self.filtered(lambda b: b.country_code == "BR")
+        bank_br.display_qr_setting = self.env.company.qr_code
+        super(ResPartnerBank, self - bank_br)._compute_display_qr_setting()
+
+    def _get_additional_data_field(self, comment):
+        """Override."""
+        if self.country_code == "BR":
+            # Only include characters allowed by the Pix spec.
+            return self._serialize(5, re.sub(r"[^a-zA-Z0-9]", "", comment))
+        return super()._get_additional_data_field(comment)
+
+    def _get_qr_code_vals_list(self, *args, **kwargs):
+        """Override. Force the amount field to always have two decimals. Uppercase the merchant name and merchant city.
+        Although not specified explicitly in the spec, not uppercasing causes errors when scanning the code."""
+        res = super()._get_qr_code_vals_list(*args, **kwargs)
+        if self.country_code == "BR":
+            res[5] = (res[5][0], float_repr(res[5][1], 2))  # amount
+            res[7] = (res[7][0], res[7][1].upper())  # merchant_name
+            res[8] = (res[8][0], res[8][1].upper())  # merchant_city
+        return res
+
+    def _get_merchant_account_info(self):
+        """Override."""
+        if self.country_code == "BR":
+            merchant_account_info_data = (
+                (0, "br.gov.bcb.pix"),  # GUI
+                (1, self.proxy_value),  # key
+            )
+            return 26, "".join(self._serialize(*val) for val in merchant_account_info_data)
+
+        return super()._get_merchant_account_info()
+
+    def _get_error_messages_for_qr(self, qr_method, debtor_partner, currency):
+        """Override."""
+        if qr_method == "emv_qr" and self.country_code == "BR":
+            if currency.name != "BRL":
+                return _("Can't generate a Pix QR code with a currency other than BRL.")
+
+        return super()._get_error_messages_for_qr(qr_method, debtor_partner, currency)
+
+    def _check_for_qr_code_errors(
+        self, qr_method, amount, currency, debtor_partner, free_communication, structured_communication
+    ):
+        """Override."""
+        if qr_method == "emv_qr" and self.country_code == "BR" and not self.proxy_type.startswith("br_"):
```

Comment:
Do not we have this logic somewhere already?

Reply:
I'm following the same pattern [as in `l10n_hk`](https://github.com/odoo/odoo/blob/14221a596b5e7ae521268db3ad424fda4bf509ea/addons/l10n_hk/models/res_bank.py#L58-L70). This override is needed because without it the accounting module [will always generate a Pix code](https://github.com/odoo/odoo/blob/cfa987933c6f91573e8ead8428c2beb531e73f4b/addons/account/models/res_partner_bank.py#L133) even if the proxy type is e.g. None.